fix(workflow): comply with branch protection by using PR-based tag updates

KillianGolds opened the pull request:
## Description

This PR updates the release automation workflow to be fully idempotent and improves the release process with enhanced safety, visibility, and reliability features. [RHOAIENG-31952](https://issues.redhat.com/browse/RHOAIENG-31952)

**Key Changes:**

1. **Idempotent Workflow Design** [Example](https://github.com/KillianGolds/odh-model-controller/actions/runs/19679689879)
   - Added tag existence checks to prevent duplicate tag creation
   - Added release existence checks to skip duplicate releases
   - Implemented branch reuse logic for existing branches
   - Added PR existence detection to reuse existing PRs instead of creating duplicates
   - File change detection for graceful no-op behavior when files already have target version

2. **Enhanced Workflow Summary** [Example](https://github.com/KillianGolds/odh-model-controller/actions/runs/19679642752)
   - Added comprehensive workflow summaries showing all actions taken
   - Four summary states: Tag Mismatch Error, No Changes Detected, Dry Run Preview, and Workflow Completed
   - Clickable links to all created resources (tags, releases, PRs, file changes)
   - File diff preview in dry-run mode to review exact changes before execution

3. **Dry-Run Mode Improvements** [Example](https://github.com/KillianGolds/odh-model-controller/actions/runs/19679589071)
   - Dry-run now validates all steps without creating actual tags/releases/PRs
   - Shows complete preview of expected actions and PR details
   - Displays diff previews for both `config/base/params.env` and `.tekton/odh-model-controller-push.yaml`
   - Release existence check runs in all modes for better visibility
   - Enables sanity checks before running the workflow for real

4. **Branch Protection Compliance**
   - Changed from direct pushes to PR-based workflow for Tekton file updates
   - Workflow now creates a PR against the target branch instead of pushing directly
   - Ensures compliance with repository branch protection rules

5. **Early Tag Validation** [Example](https://github.com/KillianGolds/odh-model-controller/actions/runs/19679532631)
   - Validates that `tag_name` exists in the Tekton file before proceeding
   - Fails fast with clear error summary when tags don't match
   - Prevents orphan tag/release creation when wrong tag is specified
   - Shows the actual current tag in file and suggests the correct value

6. **Better Visibility** [Example](https://github.com/KillianGolds/odh-model-controller/actions/runs/19679642752)
   - Extract and display PR numbers for easy reference
   - Direct links to view PR changes, releases, and tags
   - Clear indication of which files were updated and tag progression
   - Graceful error messages when mismatched tags are provided

7. **Code Quality Improvements**
   - Removed unused `NEW_TAG` environment variable
   - Consolidated tag creation flow into 3 clear steps for better maintainability
   - Simplified conditional logic by removing redundant dry-run specific steps
   - Replaced Python tag check with bash script for consistency
   - Replaced git commands with GitHub CLI (`gh`) for branch and tag operations

**Benefits:**
- Safe re-runs without creating duplicate artifacts
- Compliance with branch protection requirements
- Safer releases with dry-run validation
- Improved developer experience with comprehensive summaries
- Early issue detection through preview mode
- Handles edge cases gracefully (mismatched tags, existing artifacts)
- Cleaner, more maintainable workflow code

## How Has This Been Tested?

**Testing Environment:**
- Tested on personal fork [KillianGolds/odh-model-controller](https://github.com/KillianGolds/odh-model-controller/tree/release-test-odh-v3.1)
- Used a mock release branch (`release-test-odh-v3.1`) with both `params.env` and Tekton file set to `odh-v3.1` to simulate real release branch conditions
- Full E2E test suite with 4 comprehensive scenarios found [here](https://github.com/KillianGolds/odh-model-controller/actions/workflows/create-release.yaml).

**Test Scenarios:**

1. **Edge Case: Non-Existent Tag Handling**
   - Ran workflow with tags that don't exist in Tekton file (`odh-v99.1-test` → `odh-v99.2-test`)
   - Verified workflow fails fast with clear "Tag Mismatch" error summary
   - Confirmed no tags, releases, or PRs were created
   - Summary shows the actual current tag and suggests the correct value
   - Example: https://github.com/KillianGolds/odh-model-controller/actions/runs/19679532631

2. **Dry-Run Mode Testing**
   - Ran workflow with dry-run enabled using correct tags (`odh-v3.1` → `odh-v3.2`)
   - Verified no tags, releases, or PRs were created
   - Confirmed workflow summary showed complete preview of expected actions
   - Validated diff display for both `config/base/params.env` and `.tekton/odh-model-controller-push.yaml`
   - Verified release existence check runs in dry-run mode
   - Example: https://github.com/KillianGolds/odh-model-controller/actions/runs/19679589071

3. **End-to-End Workflow Testing (Fresh Run)**
   - Ran workflow without dry-run mode
   - Verified tag and release creation
   - Confirmed PR was created successfully with correct title and body
   - Validated PR updates both `.tekton/odh-model-controller-push.yaml` and `config/base/params.env`
   - Validated all clickable links in workflow summary (tag, release, PR, file changes)
   - Example: https://github.com/KillianGolds/odh-model-controller/actions/runs/19679642752

4. **Idempotency Testing (Re-run)**
   - Re-ran workflow with identical inputs from Test 3
   - Verified workflow detected existing tag and skipped creation
   - Confirmed workflow detected existing release and skipped creation
   - Validated branch reuse logic worked correctly
   - Verified file change detection showed "No Changes Detected" with clear message
   - Confirmed graceful exit with no errors or duplicate artifacts
   - Example: https://github.com/KillianGolds/odh-model-controller/actions/runs/19679689879

All tests passed successfully with expected behavior. Workflow is fully idempotent and handles all edge cases gracefully.


## Merge criteria:

- [x] JIRA(s) are linked in the PR description
- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

KillianGolds committed:
feat(workflow): switch to PR-based release workflow for branch protection compliance

- Add dry-run input parameter for safe testing
- Add step-level dry-run conditionals for tag and release creation
- Change bump-odh-tag job to create PR instead of direct push
- Add tag format validation
- Add comprehensive workflow summary with clickable links
- Extract and display PR number for easy reference
- Show file diff in dry-run mode

Signed-off-by: Killian Golds <kgolds@redhat.com>

rh-pre-commit.version: 2.3.2
rh-pre-commit.check-secrets: ENABLED

KillianGolds committed:
fix: use needs context for job output reference

- Change steps.fetch-tag.outputs.old_tag to needs.fetch-tag.outputs.old_tag
- fetch-tag is a separate job, not a step in the current job

Signed-off-by: Killian Golds <kgolds@redhat.com>

rh-pre-commit.version: 2.3.2
rh-pre-commit.check-secrets: ENABLED

KillianGolds committed:
test: set tag to odh-v3.99 for workflow testing

Signed-off-by: Killian Golds <kgolds@redhat.com>

rh-pre-commit.version: 2.3.2
rh-pre-commit.check-secrets: ENABLED

KillianGolds committed:
revert: restore tag to odh-v3.1 from incubating branch

Signed-off-by: Killian Golds <kgolds@redhat.com>

rh-pre-commit.version: 2.3.2
rh-pre-commit.check-secrets: ENABLED

brettmthompson commented on .github/workflows/create-release.yaml:
```diff
@@ -111,43 +123,171 @@ jobs:
+
+          echo "::notice::Tag validation passed"
+
+      - name: Create branch for ODH tag update
```
It would be good to cleanup any created branches on workflow failures.

KillianGolds replied:
This comment here is relevant to this discussion I believe: [HERE](https://github.com/opendatahub-io/odh-model-controller/pull/575/files/c715f8c4ffaea00a2675ca12f513cd49bbcfdc5f#r2454819835)

KillianGolds committed:
fix(workflow): move tag validation to early workflow stage

- Moved tag format validation from bump-odh-tag to validate-and-fetch-tag job
- Renamed fetch-tag job to validate-and-fetch-tag for clarity
- Added dry-run mode visibility at workflow start
- Ensured validation fails before creating any tags or releases

Addresses review feedback to validate inputs upfront, preventing
partial workflow execution when invalid tags are provided.

Signed-off-by: Killian Golds <kgolds@redhat.com>

rh-pre-commit.version: 2.3.2
rh-pre-commit.check-secrets: ENABLED

KillianGolds committed:
fix(workflow): remove unused param_env output

Removed orphaned output definition referencing non-existent step 'read_params_env'.
This output was unused across the workflow and caused invalid context access warnings.

Signed-off-by: Killian Golds <kgolds@redhat.com>

rh-pre-commit.version: 2.3.2
rh-pre-commit.check-secrets: ENABLED

KillianGolds committed:
refactor(workflow): use gh CLI JSON output for PR extraction

Replaced regex-based PR number extraction with gh CLI’s --json flag for
structured output and more reliable parsing. This implements CodeRabbit’s
suggestion to use JSON with jq instead of grep-based pattern matching.

Signed-off-by: Killian Golds <kgolds@redhat.com>

rh-pre-commit.version: 2.3.2
rh-pre-commit.check-secrets: ENABLED

KillianGolds committed:
fix(workflow): correct Python script indentation in YAML

Signed-off-by: Killian Golds <kgolds@redhat.com>

rh-pre-commit.version: 2.3.2
rh-pre-commit.check-secrets: ENABLED

KillianGolds committed:
fix(workflow): initialize PR outputs before early exits

Signed-off-by: Killian Golds <kgolds@redhat.com>

rh-pre-commit.version: 2.3.2
rh-pre-commit.check-secrets: ENABLED

KillianGolds committed:
fix(workflow): add GitHub token to update-params-env checkout

Signed-off-by: Killian Golds <kgolds@redhat.com>

rh-pre-commit.version: 2.3.2
rh-pre-commit.check-secrets: ENABLED

KillianGolds committed:
fix(workflow): correct typo in tag validation message

Signed-off-by: Killian Golds <kgolds@redhat.com>

rh-pre-commit.version: 2.3.2
rh-pre-commit.check-secrets: ENABLED

KillianGolds committed:
feat(workflow): make release workflow idempotent

Made the workflow safe to re-run after failures by checking for existing
resources before creating them. This enables recovery from partial failures
without requiring manual cleanup or causing errors.

Changes:
- Tag creation: check if tag exists and skip creation if present
- Release creation: check if release exists and skip creation if present
- Branch creation: check local and remote, reuse existing branch if found
- PR creation: check if PR exists for the branch and reuse instead of creating duplicates
- Added job output for tag_exists to coordinate skip logic across jobs
- Displayed current file version when no changes are detected for clarity

Signed-off-by: Killian Golds <kgolds@redhat.com>

rh-pre-commit.version: 2.3.2
rh-pre-commit.check-secrets: ENABLED

KillianGolds committed:
fix(workflow): add GitHub token to bump-odh-tag job

Added GH_TOKEN environment variable to the bump-odh-tag job to ensure
gh CLI commands (gh pr list, gh pr create) authenticate properly.

Signed-off-by: Killian Golds <kgolds@redhat.com>

rh-pre-commit.version: 2.3.2
rh-pre-commit.check-secrets: ENABLED

KillianGolds committed:
feat(workflow): make release workflow idempotent

Add full idempotency and improved reporting:

- Skip tag and release creation if they already exist
- Reuse existing branches and PRs for repeated runs
- Detect and skip when no file changes are present
- Consolidate params.env updates into fetch-tag job
- Enhance dry-run mode with diff previews
- Improve workflow summary with detailed status output

Signed-off-by: Killian Golds <kgolds@redhat.com>

rh-pre-commit.version: 2.3.2
rh-pre-commit.check-secrets: ENABLED

KillianGolds committed:
fix(nit): remove unnecessary backslash in markdown link text

Signed-off-by: Killian Golds <kgolds@redhat.com>

rh-pre-commit.version: 2.3.2
rh-pre-commit.check-secrets: ENABLED

brettmthompson commented on .github/workflows/create-release.yaml:
```diff
@@ -1,153 +1,351 @@
+        id: print_tag
+        run: |
+          echo "Old Tag=${{ steps.get_tag.outputs.old_tag_name }}"
+          echo "NEW_TAG=${{ github.event.inputs.tag_name }}" >> $GITHUB_ENV
```
Are we using this NEW_TAG value set in the github env anywhere?

KillianGolds replied:
Resolved in: [8f6956c](https://github.com/KillianGolds/odh-model-controller/commit/8f6956c1c4ee9df2a7b353bb3f63aa3aada46ba0)

KillianGolds committed:
refactor(workflow): improve tag creation flow clarity

- Consolidate tag existence handling into single step
- Move params.env update logic to unified preparation step
- Remove redundant dry-run specific steps
- Show git diff output in all modes for better visibility
- Simplify conditional logic for easier maintenance

Per review feedback: https://github.com/opendatahub-io/odh-model-controller/pull/575#discussion_r2475277219

Signed-off-by: Killian Golds <kgolds@redhat.com>

rh-pre-commit.version: 2.3.2
rh-pre-commit.check-secrets: ENABLED

KillianGolds committed:
refactor(workflow): remove unused NEW_TAG environment variable

- Remove unused NEW_TAG env var that was never referenced
- Improve step name from 'print tag' to 'Print tag information'
- Remove unused step ID 'print_tag'
- Keep useful logging for previous/new tags and branch name

Per review feedback: https://github.com/opendatahub-io/odh-model-controller/pull/575#discussion_r2475296014

Signed-off-by: Killian Golds <kgolds@redhat.com>

rh-pre-commit.version: 2.3.2
rh-pre-commit.check-secrets: ENABLED

KillianGolds committed:
refactor(workflow): run release existence check in all modes

- Remove dry-run conditional from 'Check if release exists' step
- Update dry-run message to reflect actual release existence status
- Improves visibility in dry-run mode by showing real state

Per review feedback: https://github.com/opendatahub-io/odh-model-controller/pull/575#discussion_r2475314660

Signed-off-by: Killian Golds <kgolds@redhat.com>

rh-pre-commit.version: 2.3.2
rh-pre-commit.check-secrets: ENABLED

KillianGolds committed:
refactor(workflow): consolidate release check logic

- Update release check message to be more specific
- Remove redundant dry-run release status step
- Check step now handles all modes with appropriate logging

Per review feedback: https://github.com/opendatahub-io/odh-model-controller/pull/575#discussion_r2475324859

Signed-off-by: Killian Golds <kgolds@redhat.com>

rh-pre-commit.version: 2.3.2
rh-pre-commit.check-secrets: ENABLED

spolti commented on .github/workflows/create-release.yaml:
```diff
@@ -1,153 +1,345 @@
+        run: |
+          if gh release view "${{ github.event.inputs.tag_name }}" >/dev/null 2>&1; then
+            echo "release_exists=true" >> $GITHUB_OUTPUT
+            echo "::warning::Release already exists for tag ${{ github.event.inputs.tag_name }}, skipping creation"
```
shouldn't we abort if the release already exists?

Also I think we should clean anything that was created so far (if any)

spolti replied:
I am more concerned about possible garbage left behind by failures.

KillianGolds committed:
refactor(workflow): remove unnecessary packages write permission

The packages:write permission is not needed for this workflow.
Only contents:write (for tags/releases) and pull-requests:write
(for creating PRs) are required.

Per review feedback: https://github.com/opendatahub-io/odh-model-controller/pull/575#discussion_r2476678923

Signed-off-by: Killian Golds <kgolds@redhat.com>

rh-pre-commit.version: 2.3.2
rh-pre-commit.check-secrets: ENABLED

KillianGolds committed:
refactor(workflow): replace Python tag check with bash script

Replace Python-based tag existence check with simpler bash script
for better consistency with the rest of the workflow and reduced
dependencies.

Per review feedback: https://github.com/opendatahub-io/odh-model-controller/pull/575#discussion_r2476704982

Signed-off-by: Killian Golds <kgolds@redhat.com>

rh-pre-commit.version: 2.3.2
rh-pre-commit.check-secrets: ENABLED

KillianGolds committed:
refactor(workflow): use gh CLI for branch and tag operations

- Replace git ls-remote with gh release list for getting latest tag
- Replace git show-ref/ls-remote with gh api for branch existence check
- Improves consistency as workflow already uses gh for PR operations

Signed-off-by: Killian Golds <kgolds@redhat.com>

rh-pre-commit.version: 2.3.2
rh-pre-commit.check-secrets: ENABLED

KillianGolds commented on .github/workflows/create-release.yaml:
```diff
@@ -1,153 +1,345 @@
+            PR_BODY="This PR updates the ODH release tag in the Konflux build configuration from \`${{ github.event.inputs.tag_name }}\` to \`${{ github.event.inputs.next_odh_tag }}\` for the next development cycle.
+
+            **Changes:**
+            - Updated \`.tekton/odh-model-controller-push.yaml\` with the new ODH tag
```
Thanks for the question @spolti! Looking at the workflow, the params.env update happens in the first job (fetch-tag) at lines 98-111. It updates the image reference from `:fast` to the release tag (e.g., `:odh-v2.35`) and commits that change as part of creating the tag itself.

The bump-odh-tag job (which creates the PR) only updates `.tekton/odh-model-controller-push.yaml` with the next development tag for the Konflux builds.

I could be misunderstanding the intent though - do you think we should also be updating params.env in the PR to point to the next tag as well?

KillianGolds committed:
fix(workflow): update params.env in PR to keep files in sync

The bump-odh-tag job now also updates config/base/params.env with the
next development tag, ensuring both params.env and the Tekton file
stay in sync as part of the same release PR.

Signed-off-by: Killian Golds <kgolds@redhat.com>

rh-pre-commit.version: 2.3.2
rh-pre-commit.check-secrets: ENABLED

KillianGolds committed:
fix(workflow): update params.env in PR to keep files in sync

The bump-odh-tag job now also updates config/base/params.env with the
next development tag, ensuring both params.env and the Tekton file
stay in sync as part of the same release PR.

Signed-off-by: Killian Golds <kgolds@redhat.com>

rh-pre-commit.version: 2.3.2
rh-pre-commit.check-secrets: ENABLED

KillianGolds committed:
fix(workflow): validate tag exists in Tekton file before creating release

Add early validation to check if tag_name exists in the Tekton file
before proceeding with tag/release creation. This prevents creating
orphan tags/releases when the specified tag doesn't match the current
file state.

Signed-off-by: Killian Golds <kgolds@redhat.com>

rh-pre-commit.version: 2.3.2
rh-pre-commit.check-secrets: ENABLED

KillianGolds committed:
fix(workflow): add summary output for tag mismatch failures

When the workflow fails due to tag_name not matching the current tag
in the Tekton file, a summary is now displayed showing:
- The requested tag vs actual tag in file
- Resolution steps to fix the issue

Signed-off-by: Killian Golds <kgolds@redhat.com>

rh-pre-commit.version: 2.3.2
rh-pre-commit.check-secrets: ENABLED

mholder6 reviewed (approved):
/lgtm
